File: domain/Knowledge.kt
package it.notes.ecosystem.domain

import java.time.LocalDate
import java.util.UUID

data class NoteLink(val id: String, val label: String, val start: Int, val end: Int)
data class NoteHeading(val title: String, val level: Int, val offset: Int)

/** Local links are plain Markdown, with stable IDs, never title-based identities. */
object Knowledge {
    private val uuid = "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"
    private val target = Regex("^notes://note/($uuid)$")
    private val link = Regex("(?<!!)\\[((?:\\\\.|[^\\]\\r\\n])*)]\\(notes://note/($uuid)\\)")
    fun targetId(uri: String): String? = target.matchEntire(uri)?.groupValues?.get(1)?.lowercase()
    fun url(id: String): String {
        require(Regex("^$uuid$").matches(id)) { "Identificatore della nota non valido." }
        return "notes://note/${id.lowercase()}"
    }
    fun insert(text: String, start: Int, end: Int, id: String, label: String): MarkdownEdit {
        val a = minOf(start, end).coerceIn(0, text.length)
        val b = maxOf(start, end).coerceIn(a, text.length)
        val safe = label.take(300).replace("\\", "\\\\").replace("[", "\\[").replace("]", "\\]")
            .replace('\n', ' ').replace('\r', ' ').ifBlank { "Nota" }
        val value = "[$safe](${url(id)})"
        return MarkdownEdit(text.replaceRange(a, b, value), a + value.length, a + value.length)
    }
    // Preserve offsets while excluding fenced/indented code and inline code spans.
    internal fun visible(text: String): BooleanArray {
        val result = BooleanArray(text.length) { true }
        var fence: Char? = null; var fenceSize = 0; var pos = 0
        for (line in text.split('\n')) {
            val raw = line.removeSuffix("\r")
            val match = Regex("^ {0,3}(`{3,}|~{3,})(.*)$").matchEntire(raw)
            var hidden = fence != null || raw.startsWith("    ") || raw.startsWith("\t")
            if (match != null) {
                val token = match.groupValues[1]; val tail = match.groupValues[2]
                if (fence == null && !(token[0] == '`' && '`' in tail)) {
                    fence = token[0]; fenceSize = token.length; hidden = true
                } else if (fence == token[0] && token.length >= fenceSize && tail.isBlank()) {
                    fence = null; hidden = true
                }
            }
            if (hidden) for (i in pos until minOf(text.length, pos + line.length)) result[i] = false
            pos += line.length + 1
        }
        val ticks = Regex("`+").findAll(text).filter { result[it.range.first] }.toList()
        val next = IntArray(ticks.size) { -1 }
        val latest = mutableMapOf<Int,Int>()
        for (n in ticks.indices.reversed()) {
            next[n] = latest[ticks[n].value.length] ?: -1
            latest[ticks[n].value.length] = n
        }
        val hiddenPrefix = IntArray(text.length + 1)
        for (n in text.indices) hiddenPrefix[n+1] = hiddenPrefix[n] + if(result[n]) 0 else 1
        var i = 0
        while(i < ticks.size) {
            val j = next[i]
            if(j >= 0 && hiddenPrefix[ticks[j].range.last+1] == hiddenPrefix[ticks[i].range.first]) {
                for(n in ticks[i].range.first..ticks[j].range.last) result[n] = false
                i = j + 1
            } else i++
        }
        return result
    }
    internal fun escaped(text: String, offset: Int): Boolean {
        var p = offset - 1; var count = 0
        while (p >= 0 && text[p--] == '\\') count++
        return count % 2 == 1
    }
    fun links(text: String): List<NoteLink> {
        val mask = visible(text)
        return link.findAll(text).filter { !escaped(text, it.range.first) && it.range.all { p -> mask[p] } }.map {
            NoteLink(it.groupValues[2].lowercase(), it.groupValues[1].replace(Regex("\\\\(.)"), "$1"), it.range.first, it.range.last + 1)
        }.toList()
    }
    fun remap(text: String, ids: Map<String, String>): String {
        var result = text
        links(text).asReversed().forEach { ref ->
            ids[ref.id]?.let { replacement ->
                val old = result.substring(ref.start, ref.end)
                val start = old.lastIndexOf("notes://note/")
                result = result.replaceRange(ref.start + start, ref.end - 1, url(replacement))
            }
        }
        return result
    }
    fun headings(text: String): List<NoteHeading> {
        val mask = visible(text); var offset = 0
        return text.split('\n').mapNotNull { line ->
            val start = offset; offset += line.length + 1
            val match = Regex("^ {0,3}(#{1,6})(?:[ \\t]+|$)(.*)$").matchEntire(line.removeSuffix("\r"))
            if (match == null || start >= mask.size || !mask[start]) null else
                NoteHeading(match.groupValues[2].replace(Regex("[ \\t]+#+[ \\t]*$"), "").trim(), match.groupValues[1].length, start)
        }
    }
}

object LiteralSearch {
    fun ranges(text: String, query: String, ignoreCase: Boolean = true): List<IntRange> {
        if (query.isEmpty()) return emptyList()
        val result = mutableListOf<IntRange>(); var from = 0
        while (from <= text.length - query.length) {
            val index = text.indexOf(query, from, ignoreCase)
            if (index < 0) break
            result += index until index + query.length; from = index + query.length
        }
        return result
    }
    fun replaceAll(text: String, query: String, replacement: String, ignoreCase: Boolean = true): MarkdownEdit {
        val matches = ranges(text, query, ignoreCase)
        val length = text.length.toLong() + matches.size.toLong() * (replacement.length.toLong() - query.length)
        require(length <= 200_000) { "Il risultato supera 200.000 caratteri. Riduci la sostituzione." }
        val result = buildString {
            var last = 0
            for (range in matches) { append(text, last, range.first); append(replacement); last = range.last + 1 }
            append(text, last, text.length)
        }
        return MarkdownEdit(result, 0, 0)
    }
}

data class PageTemplate(val key: String, val name: String, val title: String, val body: String)
object PageTemplates {
    val all = listOf(
        PageTemplate("meeting", "Riunione", "Riunione", "# Obiettivo\n\n## Appunti\n\n## Decisioni\n\n## Prossime azioni\n- [ ] "),
        PageTemplate("project", "Progetto", "Nuovo progetto", "# Risultato desiderato\n\n## Materiali e collegamenti\n\n## Prossimi passi\n- [ ] \n\n## Decisioni\n"),
        PageTemplate("study", "Studio", "Sessione di studio", "# Argomento\n\n## Concetti chiave\n\n## Domande\n\n## Riepilogo\n"),
        PageTemplate("review", "Revisione settimanale", "Revisione settimanale", "# Cosa ho completato\n\n## Cosa ho imparato\n\n## Da riprendere\n- [ ] \n\n## Priorità della prossima settimana\n")
    )
    fun dailyId(date: LocalDate): String = UUID.nameUUIDFromBytes("notes-ecosystem:daily:$date".toByteArray(Charsets.UTF_8)).toString()
    fun daily(date: LocalDate) = PageTemplate("daily", "Diario di oggi", "Diario · $date", "# $date\n\n## Oggi conta\n- [ ] \n\n## Appunti della giornata\n\n## Una cosa da ricordare\n")
}

File: flutter_app/lib/src/domain/knowledge.dart
class NoteLink {
  const NoteLink(this.id, this.label, this.start, this.end);
  final String id;
  final String label;
  final int start;
  final int end;
}

class NoteHeading {
  const NoteHeading(this.title, this.level, this.offset);
  final String title;
  final int level;
  final int offset;
}

class MarkdownSelectionEdit {
  const MarkdownSelectionEdit(this.text, this.start, this.end);
  final String text;
  final int start;
  final int end;
}

abstract final class Knowledge {
  static const _uuid =
      r'[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}';

  static bool _validId(String value) {
    final match = RegExp(_uuid).firstMatch(value);
    return match != null && match.start == 0 && match.end == value.length;
  }

  static String? targetId(String uri) {
    const prefix = 'notes://note/';
    if (!uri.startsWith(prefix)) return null;
    final candidate = uri.substring(prefix.length);
    return _validId(candidate) ? candidate.toLowerCase() : null;
  }

  static String url(String id) {
    if (!_validId(id)) {
      throw const FormatException('Identificatore della nota non valido.');
    }
    return 'notes://note/${id.toLowerCase()}';
  }

  static MarkdownSelectionEdit insert(
    String text,
    int start,
    int end,
    String id,
    String label,
  ) {
    final a = (start < end ? start : end).clamp(0, text.length).toInt();
    final b = (start < end ? end : start).clamp(a, text.length).toInt();
    var safe = label.length > 300 ? label.substring(0, 300) : label;
    safe = safe
        .replaceAll('\\', '\\\\')
        .replaceAll('[', '\\[')
        .replaceAll(']', '\\]')
        .replaceAll('\n', ' ')
        .replaceAll('\r', ' ')
        .trim();
    if (safe.isEmpty) safe = 'Nota';
    final value = '[$safe](${url(id)})';
    final next = text.replaceRange(a, b, value);
    return MarkdownSelectionEdit(next, a + value.length, a + value.length);
  }

  static List<NoteLink> links(String text) {
    final regex = RegExp(
      r'(?<!!)\[((?:\\.|[^\]\r\n])*)\]\(notes://note/(' + _uuid + r')\)',
    );
    final visible = _visible(text);
    return regex
        .allMatches(text)
        .where(
          (match) =>
              !_escaped(text, match.start) &&
              match.start < visible.length &&
              List<int>.generate(
                match.end - match.start,
                (index) => match.start + index,
              ).every((index) => visible[index]),
        )
        .map(
          (match) => NoteLink(
            match.group(2)!.toLowerCase(),
            (match.group(1) ?? '').replaceAllMapped(
              RegExp(r'\\(.)'),
              (m) => m.group(1) ?? '',
            ),
            match.start,
            match.end,
          ),
        )
        .toList(growable: false);
  }

  static String remap(String text, Map<String, String> ids) {
    var result = text;
    for (final ref in links(text).reversed) {
      final replacement = ids[ref.id];
      if (replacement == null) continue;
      final old = result.substring(ref.start, ref.end);
      final marker = old.lastIndexOf('notes://note/');
      if (marker < 0) continue;
      result = result.replaceRange(
        ref.start + marker,
        ref.end - 1,
        url(replacement),
      );
    }
    return result;
  }

  static List<NoteHeading> headings(String text) {
    final mask = _visible(text);
    final result = <NoteHeading>[];
    var offset = 0;
    for (final raw in text.split('\n')) {
      final line = raw.replaceFirst(RegExp(r'\r'), '');
      if (offset < mask.length && mask[offset]) {
        final match =
            RegExp(r'^ {0,3}(#{1,6})(?:[ \t]+|\z)(.*)').firstMatch(line);
        if (match != null) {
          final title = (match.group(2) ?? '')
              .replaceFirst(RegExp(r'[ \t]+#+[ \t]*'), '')
              .trim();
          result.add(NoteHeading(title, match.group(1)!.length, offset));
        }
      }
      offset += raw.length + 1;
    }
    return result;
  }

  static List<bool> _visible(String text) {
    final visible = List<bool>.filled(text.length, true);
    var fenceChar = '';
    var fenceSize = 0;
    var offset = 0;
    final fence = RegExp(r'^ {0,3}((?:\x60){3,}|~{3,})(.*)');
    for (final raw in text.split('\n')) {
      final line = raw.replaceFirst(RegExp(r'\r'), '');
      final match = fence.firstMatch(line);
      var hidden = fenceChar.isNotEmpty ||
          line.startsWith('    ') ||
          line.startsWith('\t');

      if (match != null) {
        final token = match.group(1)!;
        final tail = match.group(2) ?? '';
        final tokenChar = token[0];
        if (fenceChar.isEmpty &&
            !(tokenChar.codeUnitAt(0) == 0x60 &&
                tail.contains(String.fromCharCode(0x60)))) {
          fenceChar = tokenChar;
          fenceSize = token.length;
          hidden = true;
        } else if (fenceChar == tokenChar &&
            token.length >= fenceSize &&
            tail.trim().isEmpty) {
          fenceChar = '';
          fenceSize = 0;
          hidden = true;
        }
      }

      if (hidden) {
        final end = (offset + raw.length).clamp(0, text.length).toInt();
        for (var i = offset; i < end; i++) {
          visible[i] = false;
        }
      }
      offset += raw.length + 1;
    }

    final tickRegex = RegExp(r'(?:\x60)+');
    final ticks = tickRegex
        .allMatches(text)
        .where((match) => match.start < visible.length && visible[match.start])
        .toList();
    for (var i = 0; i < ticks.length; i++) {
      final open = ticks[i];
      for (var j = i + 1; j < ticks.length; j++) {
        final close = ticks[j];
        if (close.group(0)!.length != open.group(0)!.length) continue;
        var allVisible = true;
        for (var p = open.start; p < close.end; p++) {
          if (!visible[p]) {
            allVisible = false;
            break;
          }
        }
        if (allVisible) {
          for (var p = open.start; p < close.end; p++) {
            visible[p] = false;
          }
          i = j;
        }
        break;
      }
    }
    return visible;
  }

  static bool _escaped(String text, int offset) {
    var index = offset - 1;
    var count = 0;
    while (index >= 0 && text[index] == '\\') {
      count++;
      index--;
    }
    return count.isOdd;
  }
}

abstract final class LiteralSearch {
  static List<({int start, int end})> ranges(
    String text,
    String query, {
    bool ignoreCase = true,
  }) {
    if (query.isEmpty) return const [];
    final source = ignoreCase ? text.toLowerCase() : text;
    final needle = ignoreCase ? query.toLowerCase() : query;
    final result = <({int start, int end})>[];
    var from = 0;
    while (from <= source.length - needle.length) {
      final index = source.indexOf(needle, from);
      if (index < 0) break;
      result.add((start: index, end: index + needle.length));
      from = index + needle.length;
    }
    return result;
  }

  static MarkdownSelectionEdit replaceAll(
    String text,
    String query,
    String replacement, {
    bool ignoreCase = true,
  }) {
    final matches = ranges(text, query, ignoreCase: ignoreCase);
    final length =
        text.length + matches.length * (replacement.length - query.length);
    if (length > 200000) {
      throw const FormatException(
        'Il risultato supera 200.000 caratteri. Riduci la sostituzione.',
      );
    }
    final out = StringBuffer();
    var last = 0;
    for (final range in matches) {
      out
        ..write(text.substring(last, range.start))
        ..write(replacement);
      last = range.end;
    }
    out.write(text.substring(last));
    return MarkdownSelectionEdit(out.toString(), 0, 0);
  }
}
